simple-ipc.c

```diff
@@ -0,0 +1,185 @@
+#include "cache.h"
+#include "simple-ipc.h"
+#include "strbuf.h"
+#include "pkt-line.h"
+
+#ifdef GIT_WINDOWS_NATIVE
+static int initialize_pipe_name(const char *path, wchar_t *wpath, size_t alloc)
+{
+	int off = 0;
+	struct strbuf realpath = STRBUF_INIT;
+
+	if (!strbuf_realpath(&realpath, path, 0))
+		return error(_("could not normalize '%s'"), path);
+
+	off = swprintf(wpath, alloc, L"\\\\.\\pipe\\");
+	if (xutftowcs(wpath + off, realpath.buf, alloc - off) < 0)
+		return error(_("could not determine pipe path for '%s'"),
+			     realpath.buf);
+
+	/* Handle drive prefix */
+	if (wpath[off] && wpath[off + 1] == L':') {
+		wpath[off + 1] = L'_';
+		off += 2;
+	}
+
+	for (; wpath[off]; off++)
+		if (wpath[off] == L'/')
+			wpath[off] = L'\\';
+
+	strbuf_release(&realpath);
+	return 0;
+}
+
+struct ipc_handle_client_data {
+	struct ipc_command_listener *server;
+	HANDLE pipe;
+};
+
+static int reply(void *reply_data, const char *response, size_t len)
+{
+	int fd = *(int *)reply_data;
+
+	return write_packetized_from_buf(response, len, fd, 0);
```

Comment:
The commit message for this commit (https://github.com/microsoft/git/pull/277/commits/031698d3d5a55ab9e993ad4c7915040c08a44ac2)
says it doesn't use pkt-line routines.

Reply:
Whoops.

Reply:
I proposed a new commit message in https://github.com/microsoft/git/pull/277/commits/4e01e87166ea0487fb494dfa1302284016f8b75d (as a `squash!` because [we do not have a `reword!` yet](https://github.com/gitgitgadget/git/issues/259)):

```
    Add support for very simple Inter-Process Communication

    For the upcoming `fsmonitor--daemon` command, we need some way to send a
    command from some Git process to that daemon, and to receive an answer.
    In other words, we need to support Inter-Process Communication.

    By nature, Inter-Process Communication requires OS-specific
    implementations.

    Instead of using full-blown TCP/UDP socket communication (which comes
    with all kinds of challenges such as: restricting access permissions,
    collisions when multiple instances might want to listen on the same
    port, etc), we use named pipes on Windows, and plan on implementing a
    backend based on Unix sockets for Linux/Unix later.

    For now, we use a super-simple protocol: identified by a path into
    .git/, the daemon listens for clients, reads one command, then sends the
    answer and closes the connection.

    To avoid redundant coding patterns, we use the `pkt-line` protocol that
    we just libified to be able to use here: we really, really do not want
    our simple IPC client to take down the entire Git process with it when
    it could not talk to `fsmonitor--daemon`, but would rather instead fall
    back to a full scan of all tracked files.
```

(I adjusted the second-to-last paragraph minimally, and the last paragraph rather a lot.)